docarray/array/array/sequence_indexing_mixin.py

```diff
@@ -0,0 +1,216 @@
+        except TypeError:
+            return False
+    return False  # this is unreachable, but mypy wants it
+
+
+class IndexingSequenceMixin(Iterable[T_item]):
+    """
+    This mixin allow to extend a list into an object that can be indexed
+    a la mnumpy/pytorch.
+
+    You can index into a IndexingSequenceMixin like a numpy array or torch tensor:
+
+    .. code-block:: python
+        da[0]  # index by position
+        da[0:5:2]  # index by slice
+        da[[0, 2, 3]]  # index by list of indices
+        da[True, False, True, True, ...]  # index by boolean mask
+
+    You can delete items from a DocumentArray like a Python List
+
+    .. code-block:: python
+        del da[0]  # remove first element from DocumentArray
+        del da[0:5]  # remove elements for 0 to 5 from DocumentArray
+    """
+
+    _data: MutableSequence[T_item]
+
+    @abc.abstractmethod
+    def __init__(
+        self,
+        docs: Optional[Iterable[T_item]] = None,
+    ):
+        ...
+
+    @abc.abstractmethod
+    def __len__(self) -> int:
+        ...
+
+    @staticmethod
+    def _normalize_index_item(
+        item: Any,
+    ) -> Union[int, slice, Iterable[int], Iterable[bool], None]:
+        # basic index types
+        if item is None or isinstance(item, (int, slice, tuple, list)):
+            return item
+
+        # numpy index types
+        if _is_np_int(item):
+            return item.item()
+
+        index_has_getitem = hasattr(item, '__getitem__')
+        is_valid_bulk_index = index_has_getitem and isinstance(item, Iterable)
+        if not is_valid_bulk_index:
+            raise ValueError(f'Invalid index type {type(item)}')
+
+        if isinstance(item, np.ndarray) and (
+            item.dtype == np.bool_ or np.issubdtype(item.dtype, np.integer)
+        ):
+            return item.tolist()
+
+        # torch index types
+        torch_available = is_torch_available()
+        if torch_available:
+            import torch
+        else:
+            raise ValueError(f'Invalid index type {type(item)}')
+        allowed_torch_dtypes = [
+            torch.bool,
+            torch.int64,
+        ]
+        if isinstance(item, torch.Tensor) and (item.dtype in allowed_torch_dtypes):
+            return item.tolist()
+
+        return item
+
+    def _get_from_indices(self: T, item: Iterable[int]) -> T:
+        results = []
+        for ix in item:
+            results.append(self._data[ix])
+        return self.__class__(results)
+
+    def _set_by_indices(self: T, item: Iterable[int], value: Iterable[T_item]):
+        # here we cannot use _get_offset_to_doc() because we need to change the doc
+        # that a given offset points to, not just retrieve it.
+        # Future optimization idea: _data could be List[DocContainer], where
+        # DocContainer points to the doc. Then we could use _get_offset_to_container()
+        # to swap the doc in the container.
+        for ix, doc_to_set in zip(item, value):
```

Comment:
I think the spirit of this comment still makes sense and is true, but the reference to `_get_offset_to_doc()` is outdated

Reply:
actually I don't understand this comment

Reply:
we never had such a `_get_offset_to_doc` method

Reply:
actually with this refactor the comment maybe doesn't make sense at all, i suggest we remove it